torch/_guards.py

```diff
@@ -1140,13 +1151,24 @@ def dataclass_with_cached_hash(
 def dataclass_with_cached_hash(
     cls: type[T] | None = None, **kwargs: Any
 ) -> type[T] | Callable[[type[T]], type[T]]:
+    """
+    Decorator that creates a frozen dataclass with cached, **stable** hashing.
+
+    This decorator wraps dataclasses.dataclass and overrides __hash__ to:
+    1. Use a stable hash function that is consistent across Python processes
+       (unlike Python's built-in hash() which varies with PYTHONHASHSEED)
+    2. Cache the hash value for performance
+
+    This is important for serialization/deserialization scenarios where
+    hash consistency across processes is required.
+    """
+
     def wrap(cls_inner: type[T]) -> type[T]:
         new_cls = dataclasses.dataclass(cls_inner, **kwargs)
-        old_hash = cls_inner.__hash__
 
-        def __hash__(self) -> int:
+        def __hash__(self: Any) -> int:
             if not hasattr(self, "_hash"):
-                object.__setattr__(self, "_hash", old_hash(self))
+                object.__setattr__(self, "_hash", _stable_hash(self))
```

Comment:
I think I would kind of like to know why `__hash__` is being used for a cross-process hash. It shouldn't be; you should be using some other API in this case.